README
Net-LDAP-Gateway
================

This module provides a set of functions for packing and unpacking LDAP
messages.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2009, 2010, 2011 by Qindel Formacion y Servicios S.L.

This Perl module is free software: you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This Perl module is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this package. If not, see L<http://www.gnu.org/licenses/>.


Changes
Revision history for Perl extension Net::LDAP::Gateway.

0.03
        - packing of substrings filters was broken

0.02  Jul 11, 2011

        - documentation error for methods besed on
          _ldap_pack_result_response amended (reported by Rafael
          Porres)

0.01  Mon Jun  1 16:43:59 2009
	- original version; created by h2xs 1.23 with options
		-An Net::LDAP::Gateway



t/gen_test_data.pl
#!/usr/bin/perl

use strict;
use warnings;

use Net::LDAP::ASN qw(LDAPRequest LDAPResponse);
use Net::LDAP::Gateway;
use Data::Dumper;
use Net::LDAP::Gateway::Constant qw(:all);

$Data::Dumper::Useqq = 1;
$Data::Dumper::Purity = 1;

my @message = ( { asn1 => [ bindRequest => { version        => 3,
					     name           => 'cn=foo, o=internet',
					     authentication => { simple => 'password' },
					   }
			  ],
		  perl => [ LDAP_OP_BIND_REQUEST, { version  => 3,
						    dn       => 'cn=foo, o=internet',
						    method   => LDAP_AUTH_SIMPLE,
						    password => 'password'
						  }
			  ],
		  peek => 'cn=foo, o=internet',
		},

# 		{ asn1 => [ bindRequest => { version        => 3,
# 					     name           => 'cn=foo, o=internet',
# 					     authentication =>
# 					     { sasl => { mechanism   => 'foo',
# 							 credentials => 'credentials data'
# 						       }
# 					     }
# 					   }
# 			  ],
# 		  perl => [ LDAP_OP_BIND_REQUEST, { version          => 3,
# 						    dn               => 'cn=foo, o=internet',
# 						    method           => LDAP_AUTH_SASL,
# 						    sasl_mechanism   => 'foo',
# 						    sasl_credentials => 'credentials data'
# 						  }
# 			  ]
# 		},

		{ asn1 => [ 'unbindRequest' => {} ],
		  perl => [ LDAP_OP_UNBIND_REQUEST, {} ]
		},

		{ asn1 => [ searchRequest => { baseObject   => 'cn=bar, o=org',
					       scope        => 0,
					       timeLimit    => 300,
					       sizeLimit    => 200,
					       typesOnly    => 0,
					       derefAliases => 1,
					       filter =>
					       { and =>
						 [ { present => 'objectClass' },
						   { equalityMatch =>
						     { assertionValue => 'aval',
						       attributeDesc => 'adesc'
						     }
						   }
						 ]
					       },
					       attributes => [qw(objectClass givenName)],
					     }
			  ],
		  perl => [ LDAP_OP_SEARCH_REQUEST, { base_dn       => 'cn=bar, o=org',
						      scope         => LDAP_SCOPE_BASE_OBJECT,
						      deref_aliases => LDAP_DEREF_ALIASES_IN_SEARCHING,
						      size_limit    => 200,
						      time_limit    => 300,
						      filter        => [ LDAP_FILTER_AND,
									 [ LDAP_FILTER_PRESENT, 'objectClass'],
									 [ LDAP_FILTER_EQ, 'adesc', 'aval' ] ],
						      attributes    => [qw(objectClass givenName)]
						    }
			  ],
		  peek => 'cn=bar, o=org',
		},
		{ asn1 => [ searchRequest => { baseObject   => 'cn=bar, o=org',
					       scope        => 2,
					       derefAliases => 3,
					       sizeLimit    => 0,
					       timeLimit    => 0,
					       typesOnly    => 0,
					       filter       => { present => 'objectClass' },
					       attributes   => [],
					     }
			  ],
		  perl => [ LDAP_OP_SEARCH_REQUEST, { base_dn       => 'cn=bar, o=org',
						      scope         => LDAP_SCOPE_WHOLE_SUBTREE,
						      deref_aliases => LDAP_DEREF_ALIASES_ALWAYS,
						      filter        => [ LDAP_FILTER_PRESENT, 'objectClass'],
						    }
			  ],
		  peek => 'cn=bar, o=org'
		},
		{ asn1 => [ modifyRequest => { object => ('cn=paco,o=bar' x 500),
					       modification =>
					       [ { modification =>
						   { type => 'foo',
						     vals => ['hello']
						   },
						   operation => 2
						 }
					       ]
					     }
			  ],
		  perl => [ LDAP_OP_MODIFY_REQUEST, { dn => ('cn=paco,o=bar' x 500),
						      operations =>
						      [ { operation => LDAP_MODOP_REPLACE,
							  attribute => 'foo',
							  values => ['hello']
							}
						      ]
						    }
			  ],
		  peek => ('cn=paco,o=bar' x 500)
		},
		{ asn1 => [ modifyRequest => { object => 'cn=paco,o=bar',
					       modification =>
					       [ { modification =>
						   { type => 'foo',
						     vals => [ 'hello' ]
						   },
						   operation => 2
						 },
						 { modification => { type => 'bar',
								     vals => [ 'bye', 'really']
								   },
						   operation => 0
						 },
						 { modification => { type => 'doz',
								     vals => [ 'coz', 'muu' ]
								   },
						   operation => 1
						 },
						 { modification => { type => 'yyy',
								     vals => []
								   },
						   operation => 1
						 }
					       ]
					     }
			  ],
		  perl => [ LDAP_OP_MODIFY_REQUEST, { dn => 'cn=paco,o=bar',
						     operations =>
						     [ { operation => LDAP_MODOP_REPLACE,
							 attribute => 'foo',
							 values => [ 'hello' ],
						       },
						       { operation => LDAP_MODOP_ADD,
							 attribute => 'bar',
							 values => ['bye', 'really'],
						       },
						       { operation => LDAP_MODOP_DELETE,
							 attribute => 'doz',
							 values => ['coz', 'muu'],
						       },
						       { operation => LDAP_MODOP_DELETE,
							 attribute => 'yyy',
							 values => [],
						       }
						     ]
						   }
			 ],
		  peek => 'cn=paco,o=bar'
		},
		{ asn1 => [ addRequest => { objectName => 'cn=paco,o=bar',
					    attributes =>
					    [ { type => 'bar',
                                                vals => [ 'bye', 'really' ]
                                              }
                                            ]
					  }
			  ],
		  perl => [ LDAP_OP_ADD_REQUEST, { bar => [ 'bye', 'really' ],
						   dn => 'cn=paco,o=bar'
						 }
			  ],
		  peek => 'cn=paco,o=bar'
		},
		{ asn1 => [ delRequest => 'ou=foo,o=org' ],
		  perl => [ LDAP_OP_DELETE_REQUEST, { dn => 'ou=foo,o=org' } ],
		  peek => 'ou=foo,o=org'
		},
		{ asn1 => [ modDNRequest =>
			    { entry => 'cn=Modify Me,dc=example,dc=com',
			      deleteoldrdn => 1,
			      newSuperior => 'o=mama,o=org',
			      newrdn => 'cn=The New Me'
			    }
			  ],
		  perl => [ LDAP_OP_MODIFY_DN_REQUEST,
			    { new_superior => 'o=mama,o=org',
			      new_rdn => 'cn=The New Me',
			      delete_old_rdn => 1,
			      dn => 'cn=Modify Me,dc=example,dc=com'
			    }
			  ],
		  peek => 'cn=Modify Me,dc=example,dc=com'
		},
		{ asn1 => [ compareRequest => { entry => 'ou=foo,o=org',
						ava => { attributeDesc => 'foo',
							 assertionValue => 'koko'
						       }
					      }
			  ],
		  perl => [ LDAP_OP_COMPARE_REQUEST,
			    { dn => 'ou=foo,o=org',
			      attribute => 'foo',
			      value => 'koko'

			    }
			  ],
		  peek => 'ou=foo,o=org',
		},
		{ asn1 => [ abandonRequest => 2 ],
		  perl => [ LDAP_OP_ABANDON_REQUEST,
			    { message_id => 2 }
			  ],
		  peek => 2
		},
		{ asn1 => [ abandonRequest => 58675 ],
		  perl => [ LDAP_OP_ABANDON_REQUEST,
			    { message_id => 58675 }
			  ],
		  peek => 58675
		},
		{ asn1 => [ extendedReq => { requestName => '16.4.3',
					     requestValue => 'mi casa, telefono'
					   }
			  ],
		  perl => [ LDAP_OP_EXTENDED_REQUEST,
			    { oid => '16.4.3',
			      value => 'mi casa, telefono'
			    }
			  ]
		},
		{ asn1 => [ bindResponse => { resultCode => 1,
					      matchedDN => "o=foo",
					      errorMessage => "Bar",
					      serverSaslCreds => "vito",
					      referral => [ 'done', 'max' ]
					    }
			  ],
		  perl => [ LDAP_OP_BIND_RESPONSE,
			    { result => LDAP_OPERATIONS_ERROR,
			      matched_dn => 'o=foo',
			      message => 'Bar',
			      sasl_credentials => 'vito',
			      referrals => [ 'done', 'max' ]
			    }
			  ],
		  peek => LDAP_OPERATIONS_ERROR
		},
		{ asn1 => [ searchResEntry => { objectName => 'ou=bar,o=foo',
						attributes =>
						[ { type => 'moo',
						    vals => [ 'miau', 'dont' ] } ] } ],
		  perl => [ LDAP_OP_SEARCH_ENTRY_RESPONSE,
			    { dn => 'ou=bar,o=foo',
			      moo => ['miau', 'dont' ] } ],
		  peek => 'ou=bar,o=foo'
		},
		{ asn1 => [ searchResRef => [ qw(foo bar doz miaou)] ],
		  perl => [ LDAP_OP_SEARCH_REFERENCE_RESPONSE,
			    { uris => [ qw(foo bar doz miaou) ] } ],
		},
		{ asn1 => [ searchResDone => { resultCode => 2,
					       matchedDN => '',
					       errorMessage => 'Super-Coco' } ],
		  perl => [ LDAP_OP_SEARCH_DONE_RESPONSE,
			    { result => LDAP_PROTOCOL_ERROR,
			      matched_dn => '',
			      message => 'Super-Coco' } ],
		  peek => LDAP_PROTOCOL_ERROR
		},
		{ asn1 => [ extendedResp => { resultCode => 3,
					      matchedDN => 'ou=doom,o=com',
					      errorMessage => 'Tiriron',
					      referral => [ 'quo', 'vadis' ],
					      responseName => 'my name',
					      response => ('my value' x 1000),
					    } ],
		  perl => [ LDAP_OP_EXTENDED_RESPONSE,
			    { result => LDAP_TIME_LIMIT_EXCEEDED,
			      matched_dn =>  'ou=doom,o=com',
			      message => 'Tiriron',
			      referrals => [ 'quo', 'vadis' ],
			      name => 'my name',
			      value => ('my value' x 1000)} ],
		  peek => LDAP_TIME_LIMIT_EXCEEDED
		},
		# TODO:
		# - intermediate response tests
	      );

my @control = ( { asn1 => { type => '12.3.4.5' },
		  perl => { type => '12.3.4.5' } },
		{ asn1 => { type => '1.2.3.4',
			    critical => 1 },
		  perl => { type => '1.2.3.4',
			    criticality => 1 } },
		{ asn1 => { type => '2.1.2',
			    value => 'hello control' },
		  perl => { type => '2.1.2',
			    value => 'hello control' } },
		{ asn1 => { type => '3.1.2',
			    critical => 1,
			    value => 'bye control' },
		  perl => { type => '3.1.2',
			    criticality => 1,
			    value => 'bye control' } },
	      );

for my $req (@message) {
    my $msgid = int rand 100000;
    print "packing $msgid $req->{asn1}[0]\n";

    my $perl = $req->{perl};
    my $asn1 = $req->{asn1};
    unshift @$perl, $msgid;

    my $packer;
    if ($asn1->[0] =~ /^([a-z]+Res(p(onse)?)?|searchRes.*)$/) {
	# print "using response packer\n";
	$asn1 = [ protocolOp => { $asn1->[0] => $asn1->[1] } ];
	$packer = $LDAPResponse;
    }
    else {
	$packer = $LDAPRequest;
    }

    my @c = grep int(rand 1.5), @control;
    if (@c) {
	push @$asn1, controls => [map $_->{asn1}, @c];
	push @$perl, [map $_->{perl}, @c];
    }

    my $packed = $packer->encode(@$asn1, messageID => $msgid);
    $req->{packed} = $packed;
    unless (defined $packed) {
	print "error: ", $packer->error, "\n";
    }
}

open OUT, '>', 't/messages.pl' or die "unable to open messages.pl";
print OUT Data::Dumper->Dump([\@message], [qw(message)]);

__END__


pack_modify_request.c
#include "pack.h"
#include "util.h"

static void
ldap_pack_modop(SV *dest, SV *change) {
    if (change && SvOK(change)) {
	HV *hv;
	STRLEN offset1, offset2;
	if (!SvROK(change) || !(hv = (HV*)SvRV(change)) || (SvTYPE(hv) != SVt_PVHV))
	    croak("bad change description");
	
	offset1 = start_sequence(dest);
	pack_enum(dest, SvIV(hv_fetchs_def_undef(hv, "operation")));
	offset2 = start_sequence(dest);
	pack_string_utf8(dest, hv_fetchs_def_undef(hv, "attribute"));
	pack_set_of_string_utf8(dest, hv_fetchs_def_undef(hv, "values"));
	end_sequence(dest, offset2);
	end_sequence(dest, offset1);
    }
}

void
pack_modify_request_args(SV *dest, SV *dn, SV **args, I32 n) {
    STRLEN offset1, offset2;
    I32 i;
    offset1 = start_constructed(dest, ASN1_APPLICATION|ASN1_CONSTRUCTED, LDAP_OP_MODIFY_REQUEST);
    pack_string_utf8(dest, dn);
    offset2 = start_sequence(dest);
    for (i = 0; i < n; i++)
	ldap_pack_modop(dest, args[i]);
    end_sequence(dest, offset2);
    end_constructed(dest, offset1);
}

void
pack_modify_request_ref(SV *dest, HV *hv) {
    STRLEN offset1, offset2;
    SV *changes;
    offset1 = start_constructed(dest, ASN1_APPLICATION|ASN1_CONSTRUCTED, LDAP_OP_MODIFY_REQUEST);
    pack_string_utf8(dest, hv_fetchs_def_undef(hv, "dn"));
    offset2 = start_sequence(dest);
    changes = hv_fetchs_def_undef(hv, "operations");
    if (changes && SvOK(changes)) {
	AV *av;
	I32 i, len;
	if (SvROK(changes) && (av = (AV*)SvRV(changes)) && (SvTYPE(av) == SVt_PVAV)) {
	    len = av_len(av);
	    for (i = 0; i <= len; i++)
		ldap_pack_modop(dest, av_fetch_def_undef(av, i));
	}
	else
	    ldap_pack_modop(dest, changes);
    }
    end_sequence(dest, offset2);
    end_constructed(dest, offset1);
}


pack_search_request.c
#include "pack.h"
#include "util.h"

#define EHEAD "pack_search_filter: bad filter specification, usage: "

static void
pack_search_filter(SV *dest, AV *filter) {
    if (filter) {
	SV **svp;
	I32 type;
	I32 len = av_len(filter) + 1;
	svp = av_fetch(filter, 0, 0);
	if (!svp || !*svp)
	    croak(EHEAD "[$filter_type, @args]");
	type = SvIV(*svp);
	if (type == LDAP_FILTER_PRESENT) {
	    SV **svp = av_fetch(filter, 1, 0);
	    if (!svp || !*svp)
		croak(EHEAD "[LDAP_FILTER_PRESENT, $attribute]");
	    pack_raw_utf8(dest, ASN1_CONTEXT_SPECIFIC|ASN1_PRIMITIVE, type, *svp);
	}
	else {
	    STRLEN offset = start_constructed(dest, ASN1_CONTEXT_SPECIFIC|ASN1_CONSTRUCTED, type);
	    switch(type) {
	    case LDAP_FILTER_NOT:
		if (len != 2)
		    croak(EHEAD "[LDAP_FILTER_NOT, \\@child]");
		/* fall through */
	    case LDAP_FILTER_AND:
	    case LDAP_FILTER_OR:
	    {
		I32 i;
		if (len < 2)
		    croak(EHEAD "[LDAP_FILTER_AND, \\@child1, ...]");
		for (i = 1; i < len; i++) {
		    AV *child;
		    SV **svp = av_fetch(filter, i, 0);
		    if (!svp || !*svp || !SvROK(*svp) ||
			!(child = (AV*)SvRV(*svp)) ||
			(SvTYPE(child) != SVt_PVAV))
			croak(EHEAD "[LDAP_FILTER_AND, \\@child1, \\@child2, ...]");
		    pack_search_filter(dest, child);
		}
		break;
	    }
	    case LDAP_FILTER_EQ:
	    case LDAP_FILTER_GE:
	    case LDAP_FILTER_LE:
	    case LDAP_FILTER_APPROX:
	    {
		SV **attr = av_fetch(filter, 1, 0);
		SV **val = av_fetch(filter, 2, 0);
		if (!attr || !*attr || !val || !*val)
		    croak(EHEAD "[LDAP_FILTER_EQ, $attribute, $value]");
		pack_string_utf8(dest, *attr);
		pack_string_utf8(dest, *val);
		break;
	    }
	    case LDAP_FILTER_SUBSTRINGS:
	    {
		I32 i;
		SV **attr = av_fetch(filter, 1, 0);
		STRLEN offset = start_sequence(dest);
		if (!attr || !*attr || len < 4)
		    croak(EHEAD "[LDAP_FILTER_SUBSTRINGS, $attribute, $start, @middle, $end]");
		
		for (i = 2; i < len; i++) {
		    int stype;
		    SV **str = av_fetch(filter, i, 0);
		    if (!str || !*str || !SvOK(*str) || !sv_len(*str)) continue;
		    if (i == 2)
			stype = 0;
		    else if (i == len - 1)
			stype = 2;
		    else
			stype = 1;
		    pack_raw_utf8(dest, ASN1_CONTEXT_SPECIFIC|ASN1_PRIMITIVE, stype, *str);
		}
		end_sequence(dest, offset);
                break;
	    }
	    default:
		croak("bad filter type %d", (int)type);
		break;
	    }
	    end_constructed(dest, offset);
	}
    }
    else
	pack_raw_pvn(dest, ASN1_CONTEXT_SPECIFIC|ASN1_PRIMITIVE, LDAP_FILTER_PRESENT, "objectClass", 11);
}

void
pack_search_request_args(SV *dest,
			      SV *base_dn,
			      enum ldap_scope scope,
			      enum ldap_deref_aliases deref,
			      U32 size_limit,
			      U32 time_limit,
			      U32 types_only,
			      AV *filter,
			      SV *attributes) {
    STRLEN offset = start_constructed(dest, ASN1_APPLICATION|ASN1_CONSTRUCTED, LDAP_OP_SEARCH_REQUEST);
    pack_string_utf8(dest, base_dn);
    pack_enum(dest, scope);
    pack_enum(dest, deref);
    pack_int(dest, size_limit);
    pack_int(dest, time_limit);
    pack_bool(dest, (types_only ? 1 : 0));
    pack_search_filter(dest, filter);
    pack_sequence_of_string_utf8(dest, attributes);
    end_constructed(dest, offset);
}

void
pack_search_request_ref(SV *dest, HV *hv) {
    SV *sv;
    AV *filter;
    STRLEN offset = start_constructed(dest, ASN1_APPLICATION|ASN1_CONSTRUCTED, LDAP_OP_SEARCH_REQUEST);
    pack_string_utf8(dest, hv_fetchs_def_no(hv, "base_dn"));
    pack_enum(dest, SvIV(hv_fetchs_def_no(hv, "scope")));
    pack_enum(dest, SvIV(hv_fetchs_def_no(hv, "deref_aliases")));
    pack_int(dest, SvIV(hv_fetchs_def_no(hv, "size_limit")));
    pack_int(dest, SvIV(hv_fetchs_def_no(hv, "time_limit")));
    pack_bool(dest, SvTRUE(hv_fetchs_def_no(hv, "types_only")));
    sv = hv_fetchs_def_undef(hv, "filter");
    if (sv && SvOK(sv)) {
	if (!SvROK(sv) || !(filter = (AV*)SvRV(sv)) || (SvTYPE(filter) != SVt_PVAV))
	    croak("pack_search_request_ref: bad filter specification, usage "
		  "[$filter_type, @args]");
    }
    else filter = 0;
    pack_search_filter(dest, filter);
    pack_sequence_of_string_utf8(dest, hv_fetchs_def_undef(hv, "attributes"));
    end_constructed(dest, offset);
}


scan_modify_request.c
#include "scan.h"
#include "enum2sv.h"

void
scan_modify_request(const char** src, const char* max, HV *out) {
    STRLEN len;
    SV *dn = newSV(0);
    AV *changes;

    hv_stores(out, "dn", dn);
    scan_string_utf8(src, max, dn);
    scan_sequence(src, max, &len);

    if (len != max - *src)
	croak("scan_modify_request: packet too short");

    changes = newAV();
    hv_stores(out, "operations", newRV_noinc((SV*)changes));

    while (*src <  max) {
	const char *operation_max;
	HV *change;
	I32 operation;
	SV *attribute;
	AV *values;
	
	change = newHV();
	av_push(changes, newRV_noinc((SV*)change));
	scan_sequence(src, max, &len);
	operation_max = *src + len;

	scan_enum(src, operation_max, &operation);
	hv_stores(change, "operation", newSVsv(ldap_modop2sv_noinc(operation)));

	scan_sequence(src, max, &len);
	if (len != operation_max - *src)
	    croak("scan_modify_request: packet too short");

	attribute = newSV(0);
	hv_stores(change, "attribute", attribute);
	scan_string_utf8(src, operation_max, attribute);
	
	scan_set(src, operation_max, &len);
	if (len != operation_max - *src)
	    croak("scan_modify_request: packet too short");

	values = newAV();
	hv_stores(change, "values", newRV_noinc((SV*)values));
	while (*src < operation_max) {
	    SV *v = newSV(0);
	    av_push(values, v);
	    scan_string_utf8(src, operation_max, v);
	}
    }
}



lib/Net/LDAP/Gateway.pm
package Net::LDAP::Gateway;

our $VERSION = '0.02';

use strict;
use warnings;

require Exporter;
our @ISA = qw(Exporter);
our @EXPORT = qw( ldap_peek_message

		  ldap_shift_message

		  ldap_pack_bind_request
		  ldap_pack_bind_response
		  ldap_pack_unbind_request
		  ldap_pack_search_request
		  ldap_pack_search_entry_response
		  ldap_pack_search_reference_response
		  ldap_pack_search_done_response
		  ldap_pack_modify_request
		  ldap_pack_modify_response
		  ldap_pack_add_request
		  ldap_pack_add_response
		  ldap_pack_delete_request
		  ldap_pack_delete_response
		  ldap_pack_modify_dn_request
		  ldap_pack_modify_dn_response
		  ldap_pack_compare_request
		  ldap_pack_compare_response
		  ldap_pack_abandon_request
		  ldap_pack_extended_request
		  ldap_pack_extended_response
		  ldap_pack_intermediate_response

		  ldap_pack_message_ref
		  ldap_pack_bind_request_ref
		  ldap_pack_bind_response_ref
		  ldap_pack_unbind_request_ref
		  ldap_pack_search_request_ref
		  ldap_pack_search_entry_response_ref
		  ldap_pack_search_reference_response_ref
		  ldap_pack_search_done_response_ref
		  ldap_pack_modify_request_ref
		  ldap_pack_modify_response_ref
		  ldap_pack_add_request_ref
		  ldap_pack_add_response_ref
		  ldap_pack_delete_request_ref
		  ldap_pack_delete_response_ref
		  ldap_pack_modify_dn_request_ref
		  ldap_pack_modify_dn_response_ref
		  ldap_pack_compare_request_ref
		  ldap_pack_compare_response_ref
		  ldap_pack_abandon_request_ref
		  ldap_pack_extended_request_ref
		  ldap_pack_extended_response_ref
		  ldap_pack_intermediate_response_ref

		  ldap_dn_normalize
	       );

require XSLoader;
XSLoader::load('Net::LDAP::Gateway', $VERSION);


1;
__END__

=head1 NAME

Net::LDAP::Gateway - Infrastructure to build LDAP gateways

=head1 SYNOPSIS

  use Net::LDAP::Gateway;

=head1 DESCRIPTION

This module provides a set of low level functions to encode and decode
LDAP packets

=head2 EXPORT

The following functions can be imported from this module:

=over

=item $normalized = ldap_dn_normalize($dn)

returns a normalized version of the given dn.

=item $len = ldap_peek_message($buffer)

=item ($len, $msgid, $op, $more) = ldap_peek_message($buffer)

If enough data is available in buffer, this function returns the size
of the incomming packet. Otherwise it returns undef.

When called in list context, besides the packet len, if enough data is
available from the buffer, this function also returns the message ID,
the LDAP operation code and depending on the packet type, the dn or
the error code.

=item ($msgid, $op, $data) = ldap_shift_message($buffer)

Extracts and parses the next LDAP message available from $buffer.

C<$msgid> is the message id.

C<$op> is the number associated to the requested operation as defined
by the LDAP specification (symbolic names for those constants can be
imported from L<Net::LDAP::Gateway::Constant>).

C<$data> is a hash representing all the data available on the
packet as follows:

=over

=item LDAP_OP_BIND_REQUEST [0]

  $data = { version  => $protocol_version,
            dn       => $bind_dn,
            method   => $auth_method,
            %method_data
  }

=over

=item $auth_method == LDAP_AUTH_SIMPLE

  %method_data  =(password => $password)

=back

other authentication methods are currently not supported

=item LDAP_OP_UNBIND_REQUEST [2]

The unbind request does not carry any extra data:

  $data = {}

=item LDAP_OP_SEARCH_REQUEST [3]

  $data = { base_dn       => $base_dn,
            scope         => $scope,
            deref_aliases => $deref,
            size_limit    => $sl,      # optional
            time_limit    => $tl,      # optional
            types_only    => 1,        # optional
            filter        => \@filter
            attributes    => \@attr,   # optional
  }

C<$scope> can take the values C<LDAP_SCOPE_BASE_OBJECT> [0],
C<LDAP_SCOPE_SINGLE_LEVEL> [1] or C<LDAP_SCOPE_WHOLE_SUBTREE> [2].

C<$deref> can take the values C<LDAP_DEREF_ALIASES_NEVER> [0],
C<LDAP_DEREF_ALIASES_IN_SEARCHING> [1],
C<LDAP_DEREF_ALIASES_FINDING_BASE_OBJ> [2] or
C<LDAP_DEREF_ALIASES_ALWAYS> [3]

=item LDAP_OP_MODIFY_REQUEST [6]

  $data = { dn      => $dn,
            add     => $add,
            changes => \@changes
  }

  @changes = ( { operation => $op,
                 attribute => $attribute,
                 values    => \@values },
                 ...
             )

C<$op> can take the values C<LDAP_MODOP_REPLACE>, C<LDAP_MODOP_ADD> or
C<LDAP_MODOP_DELETE>.

=item LDAP_OP_ADD_REQUEST [8]

  $data = { dn     => $dn,
            $attr1 => \@values1,
            $attr2 => \@values2,
            ...
  }

=item LDAP_OP_DELETE_REQUEST [10]

  $data = { dn => $dn }

=item LDAP_OP_MODIFY_DN_REQUEST [12]

  $data = { dn             => $dn,
            new_rdn        => $new_rdn,
            delete_old_rdn => 1,          # optional
            new_superior   => $superior,  # optional

=item LDAP_OP_COMPARE_REQUEST [14]

  $data = { dn        => $dn,
            attribute => $attr,
            value     => $value }

=item LDAP_OP_ABANDON_REQUEST [16]

  $data = { message_id => $message_id }

=item LDAP_OP_EXTENDED_REQUEST [23]

  $data = { oid   => $oid,
            value => $value }

=item LDAP_OP_BIND_RESPONSE [1]

  $data = { result => $result_code,
            matched_dn => $dn,
            message => $message,
            referrals => \@referrals,        # optional
            sasl_credentials => $credentials # optional
  }

=item LDAP_OP_SEARCH_ENTRY_RESPONSE [5]

  $data = { dn => $entry_dn,
            $attr1 => \@values1,
            $attr2 => \@values2,
            ...
  }

=item LDAP_OP_SEARCH_DONE_RESPONSE [5]

=item LDAP_OP_MODIFY_RESPONSE [7]

=item LDAP_OP_ADD_RESPONSE [9]

=item LDAP_OP_DELETE_RESPONSE [11]

=item LDAP_OP_MODIFY_DN_RESPONSE [13]

=item LDAP_OP_COMPARE_RESPONSE [15]

=item LDAP_OP_EXTENDED_RESPONSE [24]

  $data = { result     => $result_code,
            matched_dn => $dn,
            message    => $message,
            referrals  => \@referrals    # optional
  }

C<$result_code> contains the status of the operation (see
L<Net::LDAP::Gateway::Constant/Error codes>).

=back

=item $msg = ldap_pack_message_ref($msgid, $op, \%data)

=item $msg = ldap_pack_bind_request_ref($msgid, \%data)

=item $msg = ldap_pack_bind_response_ref($msgid, \%data)

=item $msg = ldap_pack_unbind_request_ref($msgid, \%data)

=item $msg = ldap_pack_search_request_ref($msgid, \%data)

=item $msg = ldap_pack_search_entry_response_ref($msgid, \%data)

=item $msg = ldap_pack_search_done_response_ref($msgid, \%data)

=item $msg = ldap_pack_modify_request_ref($msgid, \%data)

=item $msg = ldap_pack_modify_response_ref($msgid, \%data)

=item $msg = ldap_pack_add_request_ref($msgid, \%data)

=item $msg = ldap_pack_add_response_ref($msgid, \%data)

=item $msg = ldap_pack_delete_request_ref($msgid, \%data)

=item $msg = ldap_pack_delete_response_ref($msgid, \%data)

=item $msg = ldap_pack_modify_dn_request_ref($msgid, \%data)

=item $msg = ldap_pack_modify_dn_response_ref($msgid, \%data)

=item $msg = ldap_pack_compare_request_ref($msgid, \%data)

=item $msg = ldap_pack_compare_response_ref($msgid, \%data)

=item $msg = ldap_pack_abandon_request_ref($msgid, \%data)

=item $msg = ldap_pack_extended_request_ref($msgid, \%data)

=item $msg = ldap_pack_extended_response_ref($msgid, \%data)

These functions take a C<$msgid> and a reference to a hash
containing the message data and return the LDAP message.

The data structured passed must be as documented on
C<ldap_shift_message>.

=item $msg = ldap_pack_bind_request($msgid, ...)

=item $msg = ldap_pack_bind_response($msgid, $result, $matched_dn, $message, \@referrals, $SASL_creds)

=item $msg = ldap_pack_unbind_request($msgid)

=item $msg = ldap_pack_search_request($msgid, $base_dn, $scope, $deref, $size_limit, $time_limit, $types_only, $filter, \@attributes)

=item $msg = ldap_pack_search_entry_response($msgid, $dn, $attr1 => \@values1, $attr2 => \@values2)

=item $msg = ldap_pack_search_done_response($msgid, $result, $matched_dn, $message, \@referrals)

=item $msg = ldap_pack_modify_request($msgid, ...)

=item $msg = ldap_pack_modify_response($msgid, $result, $matched_dn, $message, \@referrals)

=item $msg = ldap_pack_add_request($msgid, ...)

=item $msg = ldap_pack_add_response($msgid, $result, $matched_dn, $message, \@referrals)

=item $msg = ldap_pack_delete_request($msgid, ...)

=item $msg = ldap_pack_delete_response($msgid, $result, $matched_dn, $message, \@referrals)

=item $msg = ldap_pack_modify_dn_request($msgid, ...)

=item $msg = ldap_pack_modify_dn_response($msgid, $result, $matched_dn, $message, \@referrals)

=item $msg = ldap_pack_compare_request($msgid, ...)

=item $msg = ldap_pack_compare_response($msgid, $result, $matched_dn, $message, \@referrals)

=item $msg = ldap_pack_abandon_request($msgid, ...)

=item $msg = ldap_pack_extended_request($msgid, ...)

=item $msg = ldap_pack_extended_response($msgid, ...)

These functions take a C<$msgid> and a list of arguments and return
the corresponding LDAP message.

=back

=head1 TODO

=over

=item - add support for SASL authentication

=item - add support for common controls and extensions (requests welcome!)

=item - support controls in packing methods

=back

=head1 SEE ALSO

Other Perl LDAP related modules: L<Net::LDAP>, L<Net::LDAPapi>,
L<Net::LDAP::Server>.

LDAP RFCs 4511, 4513, 4515, 4517, 4519, 4510, 4512, 4514, 4516 and 4518.

=head1 AUTHOR

Salvador FandiE<ntilde>o, E<lt>sfandino@yahoo.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009, 2010, 2011 by Qindel Formacion y Servicios S.L.

This Perl module is free software: you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This Perl module is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this package. If not, see L<http://www.gnu.org/licenses/>.

=cut
